### PR TITLE
Improve Pointer Support

### PR DESCRIPTION
**Pull request by PolygonalSun**
PR for Improved Pointer Support.  This PR adds supports for additional mouse buttons and the mouse wheel for Mouse implementations and multi-touch support for touch inputs.

- [x] Win32 Mouse
- [x] UWP Mouse
- [x] MacOS Mouse
- [x] iOS Touch
- [x] Android Touch
- [x] Linux Mouse

### CODE REVIEW

File: Plugins/NativeInput/Source/Shared/NativeInput.cpp [hunk 1]
```diff
@@ -188,7 +212,7 @@ namespace Babylon::Plugins
                 {
                     return;
                 }
-            }
+            }*/
```
Comment (ryantrem) [inline Plugins/NativeInput/Source/Shared/NativeInput.cpp:215]: If we are sure block this is no longer needed, just remove it.

File: Plugins/NativeInput/Source/Shared/NativeInput.cpp [hunk 2]
```diff
@@ -111,6 +130,11 @@ namespace Babylon::Plugins
         PointerMove(MOUSE_POINTER_ID, x, y, DeviceType::Mouse);
     }
 
+    void NativeInput::Impl::MouseWheel(uint32_t wheelAxis, int32_t scrollValue)
+    {
+        PointerScroll(MOUSE_POINTER_ID, wheelAxis, scrollValue, DeviceType::Mouse);
```
Comment (ryantrem) [inline Plugins/NativeInput/Source/Shared/NativeInput.cpp:135]: Is there such thing as "pointer scroll" for anything other than a mouse?

File: Plugins/NativeInput/Source/Shared/NativeInput.cpp [hunk 3]
```diff
@@ -234,6 +258,19 @@ namespace Babylon::Plugins
...
 
+    void NativeInput::Impl::PointerScroll(uint32_t pointerId, uint32_t scrollAxis, int32_t scrollValue, DeviceType deviceType)
+    {
+        m_runtimeScheduler([pointerId, scrollAxis, scrollValue, deviceType, this]()
+            {
+                std::vector<int32_t>& deviceInputs{GetOrCreateInputMap(deviceType, pointerId, {scrollAxis})};
+                SetInputState(deviceType, pointerId, scrollAxis, scrollValue, deviceInputs, true);
+
+                m_eventDispatcher.tick(arcana::cancellation::none());
+                SetInputState(deviceType, pointerId, scrollAxis, 0, deviceInputs, true);
```
Comment (ryantrem) [inline Plugins/NativeInput/Source/Shared/NativeInput.cpp:269]: Why is this set back to 0?

File: Apps/Playground/UWP/App.cpp [hunk 4]
```diff
@@ -91,6 +91,9 @@ void App::SetWindow(CoreWindow^ window)
     window->PointerReleased +=
         ref new TypedEventHandler<CoreWindow^, PointerEventArgs^>(this, &App::OnPointerReleased);
 
+    window->PointerWheelChanged +=
+        ref new TypedEventHandler<CoreWindow ^, PointerEventArgs ^>(this, &App::OnPointerWheelChanged);
```
Comment (ryantrem) [inline Apps/Playground/UWP/App.cpp:95]:
```suggestion
        ref new TypedEventHandler<CoreWindow^, PointerEventArgs^>(this, &App::OnPointerWheelChanged);
```

File: Plugins/NativeInput/Source/Shared/NativeInput.cpp (hunk 2 above)
Comment (PolygonalSun) [reply Plugins/NativeInput/Source/Shared/NativeInput.cpp:135]: Not really.  Closest thing would be a touchpad gesture but that would still register as a mouse wheel scroll

File: Plugins/NativeInput/Source/Shared/NativeInput.cpp (hunk 1 above)
Comment (PolygonalSun) [reply Plugins/NativeInput/Source/Shared/NativeInput.cpp:215]: Good catch!  I'll remove it

File: Plugins/NativeInput/Source/Shared/NativeInput.cpp (hunk 3 above)
Comment (PolygonalSun) [reply Plugins/NativeInput/Source/Shared/NativeInput.cpp:269]: Mousewheel events don't have a "wheelup" event so if we don't zero it out once the frame is complete, anything that is polling or checking this value will continue to see this as a non-zero value and can keep applying it.  Because of this, we need to zero it out when the inputChanged callback is complete

File: Apps/Playground/UWP/App.cpp (hunk 4 above)
Comment (PolygonalSun) [reply Apps/Playground/UWP/App.cpp:95]: Removed while space

File: Apps/Playground/UWP/App.cpp [hunk 5]
```diff
@@ -206,12 +209,16 @@ void App::OnWindowClosed(CoreWindow^ sender, CoreWindowEventArgs^ args)
     Uninitialize();
 }
 
-void App::OnPointerMoved(CoreWindow^, PointerEventArgs^ args)
+void App::OnPointerMoved(CoreWindow^ sender, PointerEventArgs^ args)
```
Comment (bghgary) [inline Apps/Playground/UWP/App.cpp:212]: why add this?

```suggestion
void App::OnPointerMoved(CoreWindow^, PointerEventArgs^ args)
```

File: Apps/Playground/UWP/App.cpp [hunk 6]
```diff
@@ -206,12 +209,16 @@ void App::OnWindowClosed(CoreWindow^ sender, CoreWindowEventArgs^ args)
...
 {
     if (m_nativeInput != nullptr)
     {
         const auto& point = args->CurrentPoint->RawPosition;
+
         m_nativeInput->MouseMove(static_cast<int>(point.X), static_cast<int>(point.Y));
+        // Once move is handled, check for button changes
+        OnPointerPressed(sender, args);
+        OnPointerReleased(sender, args);
```
Comment (bghgary) [inline Apps/Playground/UWP/App.cpp:221]: Why do you need this? Maybe an explanation in the comment would be good.

File: Apps/Playground/UWP/App.cpp [hunk 7]
```diff
@@ -220,7 +227,21 @@ void App::OnPointerPressed(CoreWindow^, PointerEventArgs^ args)
...
+        if (!m_middlePressed && args->CurrentPoint->Properties->IsMiddleButtonPressed)
+        {
+            m_middlePressed = true;
+            m_nativeInput->MouseDown(Babylon::Plugins::NativeInput::MIDDLE_MOUSE_BUTTON_ID, static_cast<int>(point.X), static_cast<int>(point.Y));
+        }
+        if (!m_rightPressed && args->CurrentPoint->Properties->IsRightButtonPressed)
+        {
+            m_rightPressed = true;
+            m_nativeInput->MouseDown(Babylon::Plugins::NativeInput::RIGHT_MOUSE_BUTTON_ID, static_cast<int>(point.X), static_cast<int>(point.Y));
+        }
```
Comment (bghgary) [inline Apps/Playground/UWP/App.cpp:244]: I don't understand why you need to keep track of the state manually. Are the UWP events not correct? Can you do the same as Win32?

File: Apps/Playground/iOS/ViewController.swift [hunk 8]
```diff
@@ -1,6 +1,71 @@
...
+    private var _activeTouchIds: Array<Int> = [-1, -1, -1, -1, -1, -1, -1, -1, -1, -1]
+    
+    override func touchesBegan(_ touches: Set<UITouch>, with event: UIEvent) {
+        super.touchesBegan(touches, with: event)
+        
+        for touch in touches {
+            let deviceSlot = _activeTouchIds.firstIndex(of: -1)
+            
+            if (deviceSlot != nil) {
+                _activeTouchIds[deviceSlot!] = touch.hash
```
Comment (bghgary) [inline Apps/Playground/iOS/ViewController.swift:33]: What happens if there are more than 10 ids? Should we check this? Does swift arrays automatically check?

File: Apps/Playground/iOS/ViewController.swift [hunk 9]
```diff
@@ -1,6 +1,71 @@
...
+    public var onTouchDown: (Int32, Int32, Int32)->Void = {(pointerId: Int32, inX: Int32, inY: Int32) -> Void in
+        // Do nothing
+    }
+    // Callback for touch movement events
+    public var onTouchMove: (Int32, Int32, Int32)->Void = {(pointerId: Int32, inX: Int32, inY: Int32) -> Void in
+        // Do nothing
+    }
+    // Callback for touch up events
+    public var onTouchUp: (Int32, Int32, Int32)->Void = {(pointerId: Int32, inX: Int32, inY: Int32) -> Void in
+        // Do nothing
```
Comment (bghgary) [inline Apps/Playground/iOS/ViewController.swift:21]: Probably makes more sense for these to be private `let` variables and initialized by the constructor (`init`)

File: Apps/package-lock.json [hunk 10]
```diff
@@ -1,1854 +1,1854 @@
-{
```
Comment (bghgary) [inline Apps/package-lock.json:1]: Why is this file changed?

File: Plugins/NativeInput/Source/Shared/NativeInput.cpp [hunk 11]
```diff
@@ -176,20 +200,6 @@ namespace Babylon::Plugins
...
-                    index != POINTER_Y_INPUT_INDEX &&
-                    index != POINTER_DELTA_HORIZONTAL_INDEX &&
-                    index != POINTER_DELTA_VERTICAL_INDEX &&
-                    index != POINTER_MOVE_INDEX &&
-                    deviceInputs[index] > 0)
-                {
-                    return;
-                }
-            }
-
```
Comment (bghgary) [inline Plugins/NativeInput/Source/Shared/NativeInput.cpp:192]: Why is this not necessary anymore?

File: Apps/Playground/UWP/App.cpp (hunk 6 above)
Comment (PolygonalSun) [reply Apps/Playground/UWP/App.cpp:221]: So this was part of a complex solution that I had to implement because UWP input doesn't work the same as Win32 input.  First of all, PointerPressed was only firing for the first button pressed so no other buttons were getting picked up.  PointerReleased had the same issue but with only firing when the last button is released.  For the button state tracking, that was implemented as the only way I've found to track when a button is pressed/released so that I can properly use the correct event without firing unnecessary up/down events.

File: Apps/Playground/UWP/App.cpp (hunk 7 above)
Comment (PolygonalSun) [reply Apps/Playground/UWP/App.cpp:244]: Explained in previous comment

File: Apps/Playground/iOS/ViewController.swift (hunk 8 above)
Comment (PolygonalSun) [reply Apps/Playground/iOS/ViewController.swift:33]: If all of the entries in the `_activeTouchIds` are not -1 (all taken) the `firstIndex` should return `nil` so it will work like web and ignore additional touches

File: Plugins/NativeInput/Source/Shared/NativeInput.cpp (hunk 11 above)
Comment (PolygonalSun) [reply Plugins/NativeInput/Source/Shared/NativeInput.cpp:192]: For Mouse, we return before hitting that and for Touch, we only have one button to zero and it's already zeroed so this does nothing.

File: Apps/Playground/UWP/App.cpp (hunk 6 above)
Comment (PolygonalSun) [reply Apps/Playground/UWP/App.cpp:221]: I can flesh out the comment explaining this and I'll also remove the sender since we don't use it.

File: Apps/Playground/iOS/ViewController.swift (hunk 9 above)
Comment (PolygonalSun) [reply Apps/Playground/iOS/ViewController.swift:21]: Lemme see if I can get this changed over to initialize via the constructor

File: Apps/package-lock.json (hunk 10 above)
Comment (PolygonalSun) [reply Apps/package-lock.json:1]: Looks like it was a change included when I merged master in yesterday.  Lemme see if I can fix it

Comment (PolygonalSun) [reply Apps/package-lock.json:1]: Looks like it was a change included when I merged master in yesterday.  Lemme see if I can fix it

File: Plugins/NativeInput/Source/Shared/NativeInput.cpp (hunk 3 above)
Comment (ryantrem) [reply Plugins/NativeInput/Source/Shared/NativeInput.cpp:255]: What is "wheelup" (or "wheeldown")? I thought this was for an axis change (e.g. scroll) as opposed to click? I must be missing something.

File: Plugins/NativeInput/Source/Shared/NativeInput.cpp (hunk 2 above)
Comment (ryantrem) [reply Plugins/NativeInput/Source/Shared/NativeInput.cpp:135]: Meh I guess either way, I was just unsure since it was parameterized but I wasn't sure if we'd ever have any other kind of pointer with scrolling.

File: Apps/Playground/UWP/App.cpp (hunk 6 above)
Comment (bghgary) [reply Apps/Playground/UWP/App.cpp]: I don't think it's obvious why UWP doesn't send the events. A link to the documentation which says this would be great.

Comment (bghgary) [reply Apps/Playground/UWP/App.cpp]: Do we need to check `IsInContact`? The sample seems to check it, but I'm not sure we need it.

https://github.com/microsoft/Windows-universal-samples/blob/main/Samples/BasicInput/cpp/2-PointerPointProperties.xaml.cpp#L55-L60

File: Apps/Playground/UWP/App.h [hunk 12]
```diff
@@ -32,6 +32,7 @@ ref class App sealed : public Windows::ApplicationModel::Core::IFrameworkView
     void OnPointerMoved(Windows::UI::Core::CoreWindow^ sender, Windows::UI::Core::PointerEventArgs^ args);
     void OnPointerPressed(Windows::UI::Core::CoreWindow^ sender, Windows::UI::Core::PointerEventArgs^ args);
     void OnPointerReleased(Windows::UI::Core::CoreWindow^ sender, Windows::UI::Core::PointerEventArgs^ args);
+    void OnPointerWheelChanged(Windows::UI::Core::CoreWindow ^ sender, Windows::UI::Core::PointerEventArgs ^ args);
```
Comment (bghgary) [inline Apps/Playground/UWP/App.h:35]:
```suggestion
    void OnPointerWheelChanged(Windows::UI::Core::CoreWindow ^ sender, Windows::UI::Core::PointerEventArgs^ args);
```

File: Apps/Playground/UWP/App.h [hunk 13]
```diff
@@ -50,6 +51,9 @@ ref class App sealed : public Windows::ApplicationModel::Core::IFrameworkView
     Windows::Foundation::Collections::IVectorView<Windows::Storage::IStorageItem^>^ m_files;
     bool m_windowClosed;
     bool m_windowVisible;
+    bool m_leftPressed{false};
+    bool m_middlePressed{false};
+    bool m_rightPressed{false};
```
Comment (bghgary) [inline Apps/Playground/UWP/App.h:56]: nit: apps folder uses default VS conventions
```suggestion
    bool m_leftPressed{ false };
    bool m_middlePressed{ false };
    bool m_rightPressed{ false };
```

File: Apps/Playground/Win32/App.cpp [hunk 14]
```diff
@@ -401,15 +401,60 @@ LRESULT CALLBACK WndProc(HWND hWnd, UINT message, WPARAM wParam, LPARAM lParam)
             SetCapture(hWnd);
```
Comment (bghgary) [inline Apps/Playground/Win32/App.cpp:401]: Does SetCapture refcount? Does this work properly if multiple buttons are down?

**Review by bghgary — COMMENTED** [review]
Other than package-lock.json, everything else seems fine to me.

File: Plugins/NativeInput/Source/Shared/NativeInput.cpp (hunk 3 above)
Comment (PolygonalSun) [reply Plugins/NativeInput/Source/Shared/NativeInput.cpp:255]: What I mean by "wheelup" is just that when you trigger a wheelevent, there no defined end event.  Furthermore, because of how we handle input values and storing, we need a way to purge stale values.  Since these values should only be used via the input changed observable, we would just purge them at the next tick.  Basically, we turned wheel input into a sort of button.

File: Apps/Playground/Win32/App.cpp (hunk 14 above)
Comment (PolygonalSun) [reply Apps/Playground/Win32/App.cpp:401]: AFAIK, it does not.  I have modified the code to account for this.

File: Apps/Playground/UWP/App.h (hunk 13 above)
Comment (PolygonalSun) [reply Apps/Playground/UWP/App.h:56]: Changed

File: Apps/Playground/UWP/App.h (hunk 12 above)
Comment (PolygonalSun) [reply Apps/Playground/UWP/App.h:35]: Fixed

File: Apps/Playground/X11/App.cpp [hunk 15]
```diff
@@ -239,16 +239,46 @@ int main(int _argc, const char* const* _argv)
...
+                                    nativeInput->MouseDown(Babylon::Plugins::NativeInput::LEFT_MOUSE_BUTTON_ID, xmotion.x, xmotion.y);
+                                    break;
+                                case Button2:
+                                    nativeInput->MouseDown(Babylon::Plugins::NativeInput::MIDDLE_MOUSE_BUTTON_ID, xmotion.x, xmotion.y);
+                                    break;
+                                case Button3:
+                                    nativeInput->MouseDown(Babylon::Plugins::NativeInput::RIGHT_MOUSE_BUTTON_ID, xmotion.x, xmotion.y);
+                                    break;
+                                case Button4:
+                                    nativeInput->MouseWheel(Babylon::Plugins::NativeInput::MOUSEWHEEL_Y_ID, -120);
```
Comment (bghgary) [inline Apps/Playground/X11/App.cpp:256]: Just curious. Is 120 a specific number that matches browser behavior?

File: Apps/Playground/Win32/App.cpp [hunk 16]
```diff
@@ -398,20 +399,82 @@ LRESULT CALLBACK WndProc(HWND hWnd, UINT message, WPARAM wParam, LPARAM lParam)
...
             break;
         }
         case WM_LBUTTONUP:
         {
             if (nativeInput != nullptr)
             {
-                nativeInput->MouseUp(0, GET_X_LPARAM(lParam), GET_Y_LPARAM(lParam));
+                nativeInput->MouseUp(Babylon::Plugins::NativeInput::LEFT_MOUSE_BUTTON_ID, GET_X_LPARAM(lParam), GET_Y_LPARAM(lParam));
+            }
+            if (buttonRefCount-- == 1)
```
Comment (bghgary) [inline Apps/Playground/Win32/App.cpp:418]: It would be more symmetrical and very slightly faster if you did this. (Ditto everywhere else)
```suggestion
            if (--buttonRefCount == 0)
```

File: Apps/Playground/X11/App.cpp (hunk 15 above)
Comment (PolygonalSun) [reply Apps/Playground/X11/App.cpp:256]: That is the current value of deltaY when you trigger a wheel event in Chrome.

File: Apps/Playground/Win32/App.cpp (hunk 16 above)
Comment (PolygonalSun) [reply Apps/Playground/Win32/App.cpp]: Changed

**Review by bghgary — APPROVED** [review]
